src/ripple/core/SQLInterface.h

```diff
@@ -0,0 +1,247 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2020 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_CORE_SQLINTERFACE_H_INCLUDED
+#define RIPPLE_CORE_SQLINTERFACE_H_INCLUDED
+
+#include <ripple/app/ledger/Ledger.h>
+#include <ripple/app/main/Application.h>
+#include <ripple/app/misc/NetworkOPs.h>
+#include <ripple/app/misc/Transaction.h>
+#include <ripple/core/Config.h>
+#include <ripple/core/DatabaseCon.h>
+#include <ripple/peerfinder/impl/Store.h>
+#include <boost/filesystem.hpp>
+#include <boost/optional.hpp>
+#include <boost/variant.hpp>
+
+namespace ripple {
+
+struct SQLLedgerInfo
+{
+    boost::optional<std::string> sLedgerHash, sPrevHash, sAccountHash,
+        sTransHash;
+    boost::optional<std::uint64_t> totDrops, closingTime, prevClosingTime,
+        closeResolution, closeFlags, ledgerSeq64;
+};
+
+class SQLInterface
+{
+public:
+    enum DatabaseType : uint8_t { ledgerDb, transactionDb, lastDb };
+
+    enum TableType { ledgers, transactions, accountTransactions, lastTable };
+
+    typedef std::pair<LedgerIndex, uint8_t> DatabaseIndex;
+
+    static std::string
+    tableName(TableType type)
+    {
+        switch (type)
+        {
+            case ledgers:
+                return "Ledgers";
+            case transactions:
+                return "Transactions";
+            case accountTransactions:
+                return "AccountTransactions";
+            case lastTable:
+                static_assert(lastTable == 3);
+        }
+        return "Unknown";
+    }
+
+    struct RowsMinMax {
+        int rows;
+        int min;
+        int max;
+    };
+
+    struct Hashes {
+        uint256 ledgerHash;
+        uint256 parentHash;
+    };
+
+    static std::unique_ptr<SQLInterface>
+    init(Application& app, Config const& config);
+
+    virtual bool
+    makeLedgerDBs(
+        Config const& config,
+        DatabaseCon::Setup const& setup,
```

Comment:
`DatabaseCon` is specific to SQLite. Parameters of this type should be removed from the interface. You can probably hide the methods that need `DatabaseCon` in a derived class.

Reply:
This is in contrary with what Scott Determan wants. Please discuss with him.

Reply:
Can you please link me to the comment where Scott Determan is saying something different, and we can all discuss together?

Reply:
Here is link to his comment: [https://github.com/ripple/rippled/pull/3590#discussion_r479831051](https://github.com/ripple/rippled/pull/3590#discussion_r479831051). Please see section 2 where he recommends to use DatabaseCon::Setup as a parameter.

Reply:
I think Scott is talking about something slightly different than what I am saying. I think Scott is assuming the code needs `DatabaseCon` in some way no matter what. However, I am saying `DatabaseCon` is not always needed by the code. Reporting Mode uses Postgres instead of SQLite and we don't use `DatabaseCon`. Keeping it in the interface is just burdensome, and when I integrate reporting mode with this interface, I will have to refactor the interface to not use `DatabaseCon`.

Reply:
It would be really easy to hide this function inside of `SQLInterface::init()`

Reply:
@cjcobb23 I'm guessing @cdy20 is referring to this comment of mine: https://github.com/ripple/rippled/pull/3590#discussion_r479831051 I had some issues with the existing interface suggested changing to one that took the `DatabaseCon::Setup` directly. 

I'm glad you flagged this. We do want to support more than just soci connections, and adding soci specific functions on this interface is bad. However, even tho that live in `DatabaseCon` I'm not sure the setup struct itself is soci specific. It contains a collection of pragmas and a directory. I'm guessing when we add another interface we can lift `Setup` out of `DatabaseCon` and move it to, say `SQLInterface`. Hard to say until we integrate another interface. While I think this might change, I think it's OK as-is for now (unless you know it's badly broken for new interfaces, in which case let's deal with it now).

Also, minor point: `DatabaseCon` is soci specific, not sqlite specific. Soci does support more than just sqlite.

Reply:
I just think `DatabaseCon` is an implementation detail that doesn't belong in the interface. You are right that it's soci specific instead of SQLite. However, I think that is still an implementation detail, and that this interface should not have anything to do with soci. What I want to see from this interface is just a bunch functions to get ledger headers and transactions, with all database details hidden away in derived classes and implementation files. To me, soci is an implementation detail, that some implementations will not need, so it should be removed from the interface.

`DatabaseCon` also talks about filesystem paths. This is another thing we won't need with Postgres. It's very likely that Postgres will sit behind this interface, and will be the only other SQLInterface implementation besides SQLite, at least at first; we don't yet have plans for any other SQL interfaces besides SQLite and Postgres. Given that, I think at the very least that we should design this interface by looking at what both Postgres and SQLite need: keep what's common in the interface, and what's specific in derived classes. `DatabaseCon`, `soci`, and filesystem paths are not common, so should be moved to derived classes.

Reply:
@cjcobb23 I looks like we know these parameters when the concrete `SQLInterface` classes are created. If that's true, what do you think about passing in the `setup` and `checkpointerSetup` parameters into the constructor and storing them for when this function is called. Then we'd remove both parameters from this function.

I haven't looked into detail into doing that, but I think I like that. What do you think about something like that?

Reply:
@seelabs I would prefer `makeLedgersDBs` to not be part of the interface at all. `init` should take care of making the ledgers dbs, as well as any other setup. `makeLedgersDBs` could be a function of a derived class, but not the interface.

Reply:
@cjcobb23 I'm OK removing it from the interface.

Reply:
Done.